dali/pipeline/operator/builtin/external_source.h

```diff
@@ -84,13 +98,62 @@ class CachingList {
 
   void PushBack(std::list<T> &elm) {
     full_data_.splice(full_data_.end(), elm, elm.begin());
+    if (resurrect_prophet_) {
+      if (full_data_.size() == 1) {
+        prophet_ = full_data_.begin();
+      } else {
+        prophet_ = std::next(apprentice_);
+      }
+      resurrect_prophet_ = false;
+    }
+  }
+
+  const T &PeekProphet() {
+    if (prophet_ == full_data_.end())
+      throw std::out_of_range(
+          "Attempted to peek element that doesn't exist. Add more elements to CachingList before "
+          "calling PeekProphet. Even the prophet can't see outside the event horizon.");
+    return *prophet_;
+  }
+
+  void AdvanceProphet() {
+    if (prophet_ == full_data_.end())
+      throw std::out_of_range(
+          "Attempted to step over the last element in the list. This operation is forbidden. Add "
+          "more elements to CachingList before calling AdvanceProphet.");
+    apprentice_ = prophet_;
+    advance(prophet_, 1);
```

Comment:
What's wrong with `prophet_++`?
Also, what if our prophet (or `apprentice_`) is popped out of the list and put in the empty list? Wouldn't we start advancing in the set of empty nodes?

Reply:
Done